### ISSUE TITLE
Ability to prevent obscure keys

### ISSUE THREAD

**Issue by phillipsnick** [php-cache/issues#71]
Currently hierarchical keys are hashed with sha1 (https://github.com/php-cache/cache/blob/master/src/Hierarchy/HierarchicalCachePoolTrait.php#L78)

From what I can tell this isn't a requirement of PSR-6? 
Would it be possible to have a configurable option for this?

Alternatively, I'd extend this myself, but due to the way it's structured I can't see a simple way to extend the classes and override the `getHierarchyKey` method.

Perhaps the ability to define the `SomeCachePool` in config used by the adapter factory classes inside `getAdapter` would make this much simpler?

Some example points where this could be implemented:
- https://github.com/php-cache/adapter-bundle/blob/master/src/Factory/ArrayFactory.php#L30
- https://github.com/php-cache/adapter-bundle/blob/master/src/Factory/RedisFactory.php#L59

**Comment by Nyholm** [comment]
That is correct, it is not a requirement of PSR-6. If I remember correctly we do a sha1 on the hierarchy cache because there is an upper limit of the key length in Memcached. Since this is the internals of the cache adapter I assumed it did not matter. 

May I ask *why* you would like to prevent hashing of the keys?

**Comment by phillipsnick** [comment]
Theres a comment regarding that `Make sure we do not get awfully long (>250 chars) keys` which makes sense, appreciate my scenario is probably a bit of an edge case!

Currently using Redis in both the PHP application and a NodeJS application where the PHP side adds items to a queue using `RPUSH`. The NodeJS application works this list, but occasionally requires some data we store in cache. If it's not available then it simply moved it to the back of the queue until it's available. 

Probably a bad design/hack, but it's been built like that now and there's no time to find better alternatives :(

**Comment by Nyholm** [comment]
Does the data that NodeJS want to access need to be a hierarchy cache? Maybe you can store a cache key in a object that NodeJS do can access. 

If you want to override the `getHierarchyKey` you need to extend an adapter and then create a new Factory (if you are using the AdapterBundle). 

Im not sure there is a fix that we can make for the scenario.

**Comment by phillipsnick** [comment]
After investigating further I didn't realise that the data was run through `serialize` before being stored in cache. That's just going to make life interesting on the NodeJS side.

Think I'll have to connect to Redis separately and manage it via something else. Thanks for your help.

What are your thoughts on the ability to define the cache pool class in config on the adapter bundle? It would save having to extend the factory.

DoctrineBundle does something similar, eg https://github.com/doctrine/DoctrineBundle/blob/master/Resources/config/orm.xml#L13

**Comment by d3trax** [comment]
I have problem with sha1 keys when i using shared redis server with more that one application. 
I can't easy separate and see how many keys have each app. And clear all keys of one application by prefix.
Its good idea to have prefixed keys for each app in readis or memcache.
But its not possible easy override getHierarchyKey due https://github.com/php-cache/issues/issues/72

**Comment by Korbeil** [comment]
Since the key length is only a problem in `Memcached`, I think we should make it without hashes for common use and make an inherited version of it for `Memcached`.

Thoughts ?

**Comment by cryptiklemur** [comment]
Thanks for the detailed use case. The default hierarchy mapping remains hashed because the adapters share this implementation and some backends enforce short cache keys. The current [`HierarchicalCachePoolTrait` exposes `getHierarchyKey()` as protected](https://github.com/php-cache/hierarchical-cache/blob/4c200dbe126194e5ff5928dcce0b73612f73f0f9/HierarchicalCachePoolTrait.php#L30-L69), and [`cache/adapter-bundle` 2.0.1 supports application-defined factory services after Symfony merges extension definitions](https://github.com/php-cache/adapter-bundle/blob/7ca2cded07990c0bdfd8d7510dee70952274cda2/tests/Unit/DependencyInjection/CacheAdapterExtensionTest.php#L49-L65). An application can now subclass its pool, override the hierarchy mapping, and register its own factory without changing the safe default. I am closing this as addressed. If that extension path does not work with a current adapter, please reopen the issue with the subclass and configuration.
